### ISSUE TITLE
VST2 and VST3 folders separated?

### ISSUE THREAD

**Issue by perlindgren** [Goli4thus/linvstmanager#14]
Hi

First of all thanks for your efforts with linvstmanager, I think it is a very useful tool.

My setup (in short):
I'm using bitwig and reaper nativly under arch linux (with everything updated to latest rolling releases).

Latest version of wine (5.4).
Aside: I have built the linvst3 from source (with the flags to enable window size)
-DEMBEDRESIZE 

I have a question.
On the linux side I have my vst organized in two folders.
~/.vst (for the vst2)
~/.vst3 (for the vst3)

Not sure if that is the optimal setup, but I think its kind of a standard if you are building your own vsts (which I'm playing around with), I went with a default config.

My vst2 bridges goes under the ~/.vst/windows-bridged, and my vst2 bridges goes under the ~/.vst3/windows-bridged.

However: Using linvstmanager, there seems to be a single folder for storing all the bridges. Is there a way to separate the bridges (it seems to generate some naming conflicts when having both vst2s and vst3s of the same plugin).

Also, as I'm playing around with a custom bridge (built from source). Is there a way to experimentally apply the custom bridge to a single vst?

I ended up messing things up, so all my vst3s were marked "Mismatch".

As I'm pretty new to the whole vst bridging stuff, its not entirely clear if the you can have a custom .so bridge with the stock lin-vst3-servertrack, or if is possible to have multiple instances of lin-vst3-servertrack installed simultaneously. Maybe that is a question best answered by `osxmidi`.

Thanks in advance for any info.

**Comment by Goli4thus** [comment]
> First of all thanks for your efforts with linvstmanager, I think it is a very useful tool.

Thanks for the feedback and giving it try!

> I have a question.
> On the linux side I have my vst organized in two folders.
> ~/.vst (for the vst2)
> ~/.vst3 (for the vst3)
> 
> Not sure if that is the optimal setup, but I think its kind of a standard if you are building your own vsts (which I'm playing around with), I went with a default config.

I've got the same folders. Afaik those are somewhat of a standard for native plugins (at least _u-he_ installs it's VSTs into those two folders using the provided install script).
Apart from those folders some VSTs installed via package manager are placed into `/usr/lib/vst` and `/usr/lib/vst3`.

> 
> My vst2 bridges goes under the ~/.vst/windows-bridged, and my vst2 bridges goes under the ~/.vst3/windows-bridged.
> 
> However: Using linvstmanager, there seems to be a single folder for storing all the bridges. Is there a way to separate the bridges (it seems to generate some naming conflicts when having both vst2s and vst3s of the same plugin).

So far the basic idea has been to have a single `link folder` where all the symlinks to the various bridged VSTs are stored.
In general there is the _Renaming VST_ for those cases of identically named VSTs. Renaming the internal representation of the tracked VST allow to resolve the `Conflict` state (note: Renaming only affects the symlink in the `link folder`).

But if there's a VST that has both a VST2 and VST3 version with identical names (apart from file extension) and one adds both of these to _LinVstManager_, then it will in fact create a conflict initially.

To be honest I probably assumed that there would be no need to support that use case because most (if not all) people would just use one version (VST2 or VST3) of a VST and resolving a confict here and there would be _ok_.

Do you actually have that use case a lot (meaning setting up both versions simultaneously)?

> Also, as I'm playing around with a custom bridge (built from source). Is there a way to experimentally apply the custom bridge to a single vst?

I don't know your exact use case, but let's say you have a bunch of VST3s bridged using the bridge setup via _LinVst3_ (in _Preferences_).
Then you want to experimentally apply a custom bridge to one of those VST3s.
If you haven't yet _used up_ the _LinVst3-X_ bridge in _Preferences_, then you could _repurpose_ that one for selecting your experimental bridge and apply it to your desired VST3 via the `Change bridge` menu.

> I ended up messing things up, so all my vst3s were marked "Mismatch".

That probably happened because you changed the bridge, which was in use by your VSTs, within _Preferences_.
A _mismatch_ is easy to fix: Simply perform an _Edit -> Update_.

> As I'm pretty new to the whole vst bridging stuff, its not entirely clear if the you can have a custom .so bridge with the stock lin-vst3-servertrack, or if is possible to have multiple instances of lin-vst3-servertrack installed simultaneously. Maybe that is a question best answered by `osxmidi`.

Having a quick look at the Makefile of _LinVst3_ it seems that each install overwrites whatever version there was previously.
Being able to install multiple versions of the same bridge (i.e. _LinVst3_) probably is not the intended scenario.
Maybe one could rename `lin-vst3-servertrack.exe` and `lin-vst3-servertrack.exe.so` after having been installed to get away with having multiple versions in parallel. But I haven't tested that and it might not even work (that's a question osxmidi probably can answer more accurately).

Regarding _"if the you can have a custom .so bridge with the stock lin-vst3-servertrack"_:

I don't know what your intentions are here, but in general you should be able to make adaptions to LinVst3's code and build it afterwards. If your changes only affected the `.so` bridge and you end up with a seemingly _stock_ `lin-vst3-servertrack"_`, then that might just be the way it is.
Probably the most important part is that those two parts are the result of the same build, simply to be on the safe side regarding them fitting together and not introducing needless errors for yourself during trying them out.

**Comment by perlindgren** [comment]
> But if there's a VST that has both a VST2 and VST3 version with identical names (apart from file extension) and one adds both of these to _LinVstManager_, then it will in fact create a conflict initially.

I will try again, but when I tired I did not get it to work.

> 
> To be honest I probably assumed that there would be no need to support that use case because most (if not all) people would just use one version (VST2 or VST3) of a VST and resolving a confict here and there would be _ok_.
> 
> Do you actually have that use case a lot (meaning setting up both versions simultaneously)?

I'm helping Stage Craft with physical modeling, debug/beta test their pluigins, it was initially a vst2 but now also a vst3 (since the last few months). In the process there is a lot of back and forth between versions to ensure that the behaviour is correct/same. (The way CC values/controllers are propagated differs. The way to handle HDPI/re-sizing differs, etc. Also automation can be done sample accurate under VST3 but needs to be backwards compatible to VST2, as lots of users are still preferring VST2 - for some reason I don't know why).

Also I'm beta tester for Kiloheartz, they are currently implementing VST3 support (so basically the same story as with Stage Craft).

Currently, I'm using both native Windows, Windows under a VM, Wine native, and LinVst in order to ensure consistent behavior. (LinVst is the latest addition to the debug chain.) Once LinVst is stable enough I hope to get rid of both native windows and VM, to have everything under linux.

I'm also playing around with VST design under linux/Rust/druid, but this will take a while to get anything remotely useful.

My interest in LinVst (and linvstmanager) is primarily out of interest, learning about bridging, the differences between linux/windows etc. in order to make contributions.
> 
> > Also, as I'm playing around with a custom bridge (built from source). Is there a way to experimentally apply the custom bridge to a single vst?
> 
> I don't know your exact use case, but let's say you have a bunch of VST3s bridged using the bridge setup via _LinVst3_ (in _Preferences_).
> Then you want to experimentally apply a custom bridge to one of those VST3s.
> If you haven't yet _used up_ the _LinVst3-X_ bridge in _Preferences_, then you could _repurpose_ that one for selecting your experimental bridge and apply it to your desired VST3 via the `Change bridge` menu.

Ok, that's an idea. Maybe we could have a list/dropdown of bridges instead you can have/add any number of bridges. 
> 
> > I ended up messing things up, so all my vst3s were marked "Mismatch".
> 
> That probably happened because you changed the bridge, which was in use by your VSTs, within _Preferences_.
> A _mismatch_ is easy to fix: Simply perform an _Edit -> Update_.
> 
I'll experiment more with this.

> > As I'm pretty new to the whole vst bridging stuff, its not entirely clear if the you can have a custom .so bridge with the stock lin-vst3-servertrack, or if is possible to have multiple instances of lin-vst3-servertrack installed simultaneously. Maybe that is a question best answered by `osxmidi`.
> 
> Having a quick look at the Makefile of _LinVst3_ it seems that each install overwrites whatever version there was previously.
> Being able to install multiple versions of the same bridge (i.e. _LinVst3_) probably is not the intended scenario.

During development it would be great. As e.g. HDPI and re-size is still not implemented/under development.  In the long term, maybe not needed, but until such features are stable I think there is a need/purpose.

> Maybe one could rename `lin-vst3-servertrack.exe` and `lin-vst3-servertrack.exe.so` after having been installed to get away with having multiple versions in parallel. But I haven't tested that and it might not even work (that's a question osxmidi probably can answer more accurately).
> 
> Regarding _"if the you can have a custom .so bridge with the stock lin-vst3-servertrack"_:
> 
> I don't know what your intentions are here, but in general you should be able to make adaptions to LinVst3's code and build it afterwards. If your changes only affected the `.so` bridge and you end up with a seemingly _stock_ `lin-vst3-servertrack"_`, then that might just be the way it is.
> Probably the most important part is that those two parts are the result of the same build, simply to be on the safe side regarding them fitting together and not introducing needless errors for yourself during trying them out.

Yes, I haven't yet looked at the code/implementation deep enough to really understand the inner workings of the bridge, maybe we could meet up at discord/zulip/riot/telegram or some other chat to discuss interactively.

**Comment by Goli4thus** [comment]
> I will try again, but when I tired I did not get it to work.

Well, it will only work if you resolve the confict by doing `Edit -> Rename VST` regarding one of those VSTs. Otherwise the confict will persists and it can't enable that VST.

Anyway: If you still can't get it to work after trying that out, then I'll need a step by step description of what you are doing, so I can try to reproduce it.

> > To be honest I probably assumed that there would be no need to support that use case because most (if not all) people would just use one version (VST2 or VST3) of a VST and resolving a confict here and there would be _ok_.
> > Do you actually have that use case a lot (meaning setting up both versions simultaneously)?
> 
> I'm helping Stage Craft with physical modeling, debug/beta test their pluigins, it was initially a vst2 but now also a vst3 (since the last few months). In the process there is a lot of back and forth between versions to ensure that the behaviour is correct/same. (The way CC values/controllers are propagated differs. The way to handle HDPI/re-sizing differs, etc. Also automation can be done sample accurate under VST3 but needs to be backwards compatible to VST2, as lots of users are still preferring VST2 - for some reason I don't know why).
> 
> Also I'm beta tester for Kiloheartz, they are currently implementing VST3 support (so basically the same story as with Stage Craft).

Just did a quick test with bitwig demo.
Bitwig seems to behave similar to Renoise here, meaning if you rename a VST within _LinVstManager_, it will show that name within bitwig's plugin browser. So _renaming_ might actually come in quite handy here when dealing with various versions of the same VST.

> > > Also, as I'm playing around with a custom bridge (built from source). Is there a way to experimentally apply the custom bridge to a single vst?
> > 
> > 
> > I don't know your exact use case, but let's say you have a bunch of VST3s bridged using the bridge setup via _LinVst3_ (in _Preferences_).
> > Then you want to experimentally apply a custom bridge to one of those VST3s.
> > If you haven't yet _used up_ the _LinVst3-X_ bridge in _Preferences_, then you could _repurpose_ that one for selecting your experimental bridge and apply it to your desired VST3 via the `Change bridge` menu.
> 
> Ok, that's an idea. Maybe we could have a list/dropdown of bridges instead you can have/add any number of bridges.

For one, that would be quite a big change to the current implementation.
Furthermore it probably would add another layer of complexity for the end user.

The current implementation with four distinct bridges is pretty self eplanatory to the end user.
osxmidi has created four different bridge variants and all of them can be set up in _LinVstManager_.

If it would be possible to add an arbitrary number of bridges (a distinction for VST2 and VST3 would need to be made), then the average end user would probably just be confused for why there would even be a need for more bridges.

So yeah, it makes quite a difference whether one has a "user" or a "developer" mindset when looking at this application.

> > > As I'm pretty new to the whole vst bridging stuff, its not entirely clear if the you can have a custom .so bridge with the stock lin-vst3-servertrack, or if is possible to have multiple instances of lin-vst3-servertrack installed simultaneously. Maybe that is a question best answered by `osxmidi`.
> > 
> > 
> > Having a quick look at the Makefile of _LinVst3_ it seems that each install overwrites whatever version there was previously.
> > Being able to install multiple versions of the same bridge (i.e. _LinVst3_) probably is not the intended scenario.
> 
> During development it would be great. As e.g. HDPI and re-size is still not implemented/under development. In the long term, maybe not needed, but until such features are stable I think there is a need/purpose.

Well, if you think there's an actual need, then it would probably be best to talk to osxmidi directly via an issue. Maybe he can tell you how he usually switches between various versions of the same bridge during development (he is the developer after all, so chances are he has been in that situation already).


> Yes, I haven't yet looked at the code/implementation deep enough to really understand the inner workings of the bridge, maybe we could meet up at discord/zulip/riot/telegram or some other chat to discuss interactively.

TBH, I'm haven't really had a look at the inner working of those bridges either. So I can't really explain things regarding them. I'm more of a user myself and in addition developed this graphical application to help with the setup/managing part.

And if there are technical questions regarding _LinVstManager_, I'd prefer to actually keep it on github, just so others can benefit from the discussion as well.

**Comment by perlindgren** [comment]
Thanks for your feedback. I'll bring up the tech discussion with osxmidi, and we can cont. the discussion regarding the `linvstmanager` here.

Regarding the GUI I have a few improvements to suggest (but I will do that in a separate issue).

By the way, in the README.md, you may mention that there is already a aur package for `linvstmanager`, so as an end user you can just install it with your package manager.

**Comment by Goli4thus** [comment]
> By the way, in the README.md, you may mention that there is already a aur package for `linvstmanager`, so as an end user you can just install it with your package manager.

Thanks for letting me know. Totally forgot about adding that information. Readme has been updated.

> Regarding the GUI I have a few improvements to suggest (but I will do that in a separate issue).

Ok, then I'll be closing this one now.

**Comment by keybreak** [comment]
@Goli4thus 
@perlindgren 
I wanted to give some thoughts on that matter, even though i'm not user of VST3...

And this is rationally can be excluded as edge case for most users...Still i'm thinking of studios moving from Win / Mac to Linux in future (which is already very possible as far as my own experience).

For example this could be valid case for compatibility with potential client's project (mixing engineer using VST3 for most of his workflow, but user sends project which need 'fix in the mix' with VST2 version of same plugin).

Still trying to see bigger picture here:
- Most of the DAWs that support VST3 standard on Windows i think differentiate VST2 & VST3 even if they have same name...But in case of our .so files it's not - and it makes sense.
- For maximum compatibility ideal variant would be to keep original name even if it's VST2 = VST3.

I can't test it myself, but maybe something like that would work for any DAW which support VST3:
- Create separate **~/.vst** and **~/.vst3** with linvstmanager one at a time
- Create link for  **~/.vst3** as **~/.vst/.vst3**
- Scan only **~/.vst** in your DAW, should probably give you separate VST2 / VST3 plugins since VST3 path is a little different.

If that would work, probably it's a good idea to think in that direction, since we can clearly differentiate VST3 already - it can be easily linked inside separate directory inside our link directory...

@perlindgren 
Thx for your efforts in beta testing though, **kiloHearts** plugs are working really good with LinVst and Wine :+1:

**Comment by Goli4thus** [comment]
_Wall of text incoming..._

So I did some further testing with different DAWs and their scan behaviour when it comes to having identically named VST2 and VST3 files.

What I did was creating a `vst2` and `vst3` folder within the _link folder_ (~/.vst_linkFolder in my case) and created symlinks to a pair of VST2/VST3 VST divided into those two sub folders.
Looked like this (version 1):

```
.vst_linkFolder
├── vst2
│   └── MBassador.so -> /home/goli/.wines/vst_melda/drive_c/Program Files/VstPlugins/MeldaProduction/MBassador.so
└── vst3
    └── MBassador.so -> /home/goli/.wines/vst_melda/drive_c/Program Files/Common Files/VST3/MeldaProduction/MBassador.so
```

A second scenario looked like this (version 2):

```
.vst_linkFolder
├── vst2
│   └── *MBassadorVST2*.so -> /home/goli/.wines/vst_melda/drive_c/Program Files/VstPlugins/MeldaProduction/MBassador.so
└── vst3
    └── MBassador.so -> /home/goli/.wines/vst_melda/drive_c/Program Files/Common Files/VST3/MeldaProduction/MBassador.so
```

So intenionally renaming one of them so that the VST files' symlinks themselves differ in their name and not just in their path.


Then I did a test procedure for a selection of *DAW*s:
- Reaper (Full)
- Renoise (Demo)
- Bitwig (Demo)
- Ardour


The test procedure looked like this:
1. Scan `vst_linkfolder` (*version 1*) with *DAW*
2. Verify what is picked up within *DAW* (check the displayed name in FX browser)
3. Load whatever VST of the two got picked up (check the displayed name of loaded VST)
4. Scan `vst_linkfolder` (*version 2*) with *DAW*
5. Verify what is picked up within *DAW* (check the displayed name in FX browser)
6. Load whatever VST of the two got picked up (check the displayed name of loaded VST)


### Reaper
Important: Always do a "Clear cache/re-scan" within Reaper's preferences for this. Otherwise there's some real strange behaviour and Reaper seemingly isn't picking up renamd folders/files.

#### Version 1
Only one of the two is showing up as _"VST: MBassador (Melda Production)"_.
This seems to be the `vst2` one.
If I remove the `vst2` subfolder, the `vst3` one shows up with name _"VST: MBassador [vst3] (Melda Production)"_.
Once I load either VST, it shows up with the same name as was displayed in the track's FX window.

#### Version 2
Now both are picked up, shown with the following names in the FX browser and track's FX window once loaded:
- _"VST: MBassador (Melda Production)"_
- _"VST: MBassador [vst3] (Melda Production)"_

My current guess is the following:
Reaper initially somehow cares about the symlink names and skips any duplicates right away.
This would explain why the `vst3` isn't picked up in Version 1.

After that Reaper seems to actually load the VST and read its _name_ and _vendor_.
Usually Reaper detects VST3 and labels them accordingly in Windows.
In this case it might be that Reaper reads whatever LinVst3 bridge is reporting.
Maybe that's why it says _"VST: MBassador [vst3] (Melda Production)"_ and not _"VST3: MBassador (Melda Production)"_.
Or maybe Reaper actually behaves differently on linux; but I doubt that, because native u-he VST3 files actaully get labeled as "VST3i" within the FX browser.
so probably has to do with _LinVst3_ bridge.

#### Conclusion (Reaper):
Differently named symlinks and not just different folder paths are needed for both VST2 and VST3 versions to be picked up in Reaper.


### Renoise
Always did a _Rescan_ within preferences of Renoise.

#### Version 1
Only one of the two is showing up as _"MBassador"_.
This seems to be the `vst2` one.
If I remove the `vst2` subfolder, the `vst3` one shows up with name _"MBassador"_ as well.
Once I load either VST, they show up with the same name in both cases: _"VST: MBassador"_.

#### Version 2
Now both are picked up, shown with the following names in the FX browser:
- _"MBassador"_
- _"MBassadorVST2"_

On loading them into a track, the naming changes to:
- _"VST: MBassador"_
- _"VST: MBassadorVST2"_

For Renoise my guess is (as we already learned from past experiments), that Renoise really cares about the actual symlink names and does not load the name contained within the VST itself (like Reaper does).

#### Conclusion (Renoise):
Differently named symlinks and not just different folder paths are needed for both VST2 and VST3 versions to be picked up in Renoise.


### Bitwig
Added the `vst_linkFolder` to _Plug-in Locations_ within Bitwig's preferences and enabled _"Monitor this location for file-system changes"_.

#### Version 1
Bitwig actually seems to care about the VST's symlink path and shows the `vst_linkfolder` with its subfolders in a tree view within its FX browser.
Therefore even if both symlinks are named identical, they both show up as _"MBassador"_ within their respective subfolder.
Once I load each VST, they show up as:
- _"MBassador"_ (for the `vst2` one)
- _"MBassador [VST3]"_ (for the `vst2` one)

#### Version 2
Both are picked up yet again, shown with the following names in the FX browser:
- _"MBassadorVST2"_ (for the `vst2` subfolder)
- _"MBassador"_(for the `vst3` subfolder)


On loading them into a track, the naming changes to:
- _"MBassador"_ (for the `vst2` one)
- _"MBassador [VST3]"_ (for the `vst3` one)

So for Bitwig it seems that the symlink path matters and is enough for Bitwig to differentiate between identically named VSTs.
Furthermore within the FX browser the actual symlink name is being displayed and only once the VST is loaded is the name replaced with the actual embedded name within the VST itself.
This is strange however, because the FX browser already shows the _vendor_ "Melda Production" correctly.
So Bitwig somehow reads the vendor, but not the VST name at the point in time when shows VSTs in the FX browser.

#### Conclusion (Bitwig):
Differently named symlinks are not neccessary due to different paths being enough to make a difference during Bitwig's VST detection/scan.


### Ardour
Added `vst_linkFolder` to VST _Path_ within Ardour's preferences and did a "Scan for Plugins" each time.

#### Version 1
Ardour actually seems to load VSTs already for the FX browser, because both VSTs are showing up like so:
- _"MBassador" (MeldaProduction)_ (for the `vst2` one)
- _"MBassador [vst3]" (MeldaProduction)_ (for the `vst2` one)

Once I load each VST, they show up the same naming pattern.

#### Version 2
No real difference. Still both are picked up and shown with the same naming pattern.

So for Ardour it seems that the symlink naming is completely irrelevant and the only thing that matters seems to be that the VSTs actually exist on the filesystem (identically named files/symlinks can't exist within the same folder after all).
Ardours simply loads all scanned VSTs right away for the FX browser.
The drawback of this is that the scan actually takes the longest overall (loading VSTs takes time, especially when _Wine_ is involved).

#### Conclusion (Ardour):
Differently named symlinks are not neccessary. Only "valid" symlinks are needed ("valid" in the sense of they gotta exist within the scanned folder path of Ardour).



## So overall...
... I'm actually not really sure.
Those DAWs actually behave quite differently.
Ardour and Bitwig seem to allow a bit more, specifically regarding having identical symlink names.

To be on the safe side for all those DAWs the "have unique symlinks names" route seems the way to go.
But I'm not sure if auto-appending suffixes like "_vst2.so" or "vst3.so" to each symlink to tackle this VST2/VST3 problem is the right way to go.
This would end up cluttering the FX browser quite a bit in case of Bitwig and Renoise.

Furthermore I don't know if renaming symlinks actually helps with compatibiliy in regards to getting old Windows projects to run under linux. If the symlink is renamed and doesn't match the old VST2 one from Windows times, then how should the DAW understand it's the same plugin?

**Comment by perlindgren** [comment]
Thanks for the update. When you write
```"MBassador [VST3]" (for the vst2 one)```
I assume its just a typo, right, `(for the vst3 one)` it should be....

Regarding conclusion. Automatically putting vsts in subfolders vst2/vst3 (dependent on type) should not break anything, renaming might (as you mentioned old projects might not find the plugin). Implementing that would be my suggestion (if I haven't misunderstood anything that is...)

**Comment by keybreak** [comment]
Thanks for such detailed analysis!

>  But I'm not sure if auto-appending suffixes like "_vst2.so" or "vst3.so" to each symlink to tackle this VST2/VST3 problem is the right way to go.

It technically should work, but i'm sure it's not right way to go for a lot of reasons highlighted above during our previous and your latest test...

> Furthermore I don't know if renaming symlinks actually helps with compatibiliy in regards to getting old Windows projects to run under linux. If the symlink is renamed and doesn't match the old VST2 one from Windows times, then how should the DAW understand it's the same plugin?

I can speak only for Renoise, but in there with VST2 it perfectly loads projects with symlinks which have exactly same name as plugin on Windows had...And in Renoise still even if there's different name, tehcnically it's fixable on project level by unzip project and editing **xml**, but with Reaper and Bitwig i suppose not so much?
____________

So for truly being all around compatible for any Linux DAW subfolders are not an option, and renaming symlinks is also not an option...

I suppose there's not much we can do here therefore. User shouldn't use VST2 and VST3 simultaneously without consequences!! :laughing: 

____________

Renoise btw natively doesn't support VST3 as far as i know, so should be logical behavior...But not for .so apparently :smiley: 

> After that Reaper seems to actually load the VST and read its name and vendor.
Usually Reaper detects VST3 and labels them accordingly in Windows.
In this case it might be that Reaper reads whatever LinVst3 bridge is reporting.
Maybe that's why it says "VST: MBassador [vst3] (Melda Production)" and not "VST3: MBassador (Melda Production)".

~~It's definitely LinVst3 though, i had similar type of problem while testing LinVst in Renoise where it had always reported newly added vendor as **undefined**, that bug was resolved...Probably you should report this one to **LinVst3** if it's a big deal in difference for end user of VST3~~

Actually wait, it's same situation as Renoise, for them native programs on Linux apparently it's not VST2/VST3 - it's Linux only .so which totally makes sense...That's why there are no differentiation kicks in like on Windows!

**Comment by Goli4thus** [comment]
> Thanks for the update. When you write
> `"MBassador [VST3]" (for the vst2 one)`
> I assume its just a typo, right, `(for the vst3 one)` it should be....

Yeah that's a type. Corrected.

> Regarding conclusion. Automatically putting vsts in subfolders vst2/vst3 (dependent on type) should not break anything, renaming might (as you mentioned old projects might not find the plugin). Implementing that would be my suggestion (if I haven't misunderstood anything that is...)

To be honest, it was a lot of text, but if we only focus on the _Conclusion_ sections for each DAW, then:
- having different subfolders would help specifically in case of Bitwig
- but it would not solve the problem for Reaper and Renoise. They would still need differently named symlinks.

**Comment by Goli4thus** [comment]
> > But I'm not sure if auto-appending suffixes like "_vst2.so" or "vst3.so" to each symlink to tackle this VST2/VST3 problem is the right way to go.
> 
> It technically should work, but i'm sure it's not right way to go for a lot of reasons highlighted above during our previous and your latest test...

Yeah I don't like that idea either after all.

> > Furthermore I don't know if renaming symlinks actually helps with compatibiliy in regards to getting old Windows projects to run under linux. If the symlink is renamed and doesn't match the old VST2 one from Windows times, then how should the DAW understand it's the same plugin?
> 
> I can speak only for Renoise, but in there with VST2 it perfectly loads projects with symlinks which have exactly same name as plugin on Windows had...And in Renoise still even if there's different name, tehcnically it's fixable on project level by unzip project and editing **xml**, but with Reaper and Bitwig i suppose not so much?

Imo that's another reason why we should not just auto-rename symlinks.
This is not the normal use case after all.
Having differently named VSTs is though, so we should not break that usecase for any user and not break more than needed in regards to loading old projects.
The fact that if one knows how to do it, he can work around such renaming in Renoise is a plus. But I don't know about the project files of those other DAWs. Maybe something similiar is possible there as well.
But that goes beyond the scope of _LinVstManager_ is more about the intrinsics of a particular DAW.

> So for truly being all around compatible for any Linux DAW subfolders are not an option, and renaming symlinks is also not an option...
> 
> I suppose there's not much we can do here therefore. User shouldn't use VST2 and VST3 simultaneously without consequences!! laughing

Well you might be right. The consequence for the user would be that if he really has a scenario of "VST2/VST3 pair with same name", then he might simply have to make use of the renaming feature in those cases.

> Renoise btw natively doesn't support VST3 as far as i know, so should be logical behavior...But not for .so apparently smiley

Well, that's a least what it looked like. But I only tested what I wrote above. So no guarantee that it's stable if Renoise doesn't officially support VST3.

> > After that Reaper seems to actually load the VST and read its name and vendor.
> > Usually Reaper detects VST3 and labels them accordingly in Windows.
> > In this case it might be that Reaper reads whatever LinVst3 bridge is reporting.
> > Maybe that's why it says "VST: MBassador [vst3] (Melda Production)" and not "VST3: MBassador (Melda Production)".
> 
> ~It's definitely LinVst3 though, i had similar type of problem while testing LinVst in Renoise where it had always reported newly added vendor as **undefined**, that bug was resolved...Probably you should report this one to **LinVst3** if it's a big deal in difference for end user of VST3~
> 
> Actually wait, it's same situation as Renoise, for them native programs on Linux apparently it's not VST2/VST3 - it's Linux only .so which totally makes sense...That's why there are no differentiation kicks in like on Windows!

Yeah that might be a core reason for why this whole thing might not be an issue on Windows, but on Linux it is now:
Windows simply has two different file extensions for VST2 (.dll) and VST3 (.vst3), so it's easier for DAWs to keep them apart and detect them both even if the have the same filename (without the extension).

**Comment by perlindgren** [comment]
Did you come to a conclusion? (What is the best to do for vst2/vst3, separate folders, and or renaming?)

**Comment by Goli4thus** [comment]
> Did you come to a conclusion? (What is the best to do for vst2/vst3, separate folders, and or renaming?)

Well, I did think about it a bit more, but haven't really come up with an optimal solution.

One thing I learned was that having _*.vst3_ on linux is actually a real thing:

https://steinbergmedia.github.io/vst3_doc/vstinterfaces/vst3loc.html

Looking specifically at section "For the Linux platform", it's clear that:

- a ".vst3" is basically a folder/archive
- the actual VST3 file itself is still a ".so" file (the native linux shared library format)

@keybreak:
I think that might also be an explanation why in my tests above Renoise was able to load a VST3 using _Linvst3_. 
Maybe Renoise just doens't officially support parsing a real ".vst3" archive.
But if provided with a ".so" file via _LinVst3_ it actually can load it.

Disclaimer: I don't really know what LinVst3 is doing. Furthermore osxmidi clearly states that _"Not all vst3 features are supported."_ on _LinVst3_ page.


Back on topic:

Overall I think this whole thing re having VST2/VST3 in parallel and therefore potentially having two identically named ".so"-files is really strongly tied to the issue of having identically named VST2 ".so"-files by different vendors (i.e. _EQ.so_ by vendor-A and vendor-B).

The solution back then was the introduction of the _Rename_ feature, which allowed making the VST's name a DAW "sees" unique.
Of course, the core problem back then was about solving the "how to have identically named VST2 files in the same _link folder_.

But overall _Rename_ was (and still is) a solution that worked for all of those DAWs.
Plus: _Rename_ works just as well for solving that VST2/VST3 in parallel problem.


So for the time being, I think _Rename_ is the answer.


@perlindgren 
About how many such scenrios are we talking in your case?
10? 20? 50? How many VST2/VST3 in parallel cases do you realistically have?

**Comment by perlindgren** [comment]
"About how many such scenrios are we talking in your case?
10? 20? 50? How many VST2/VST3 in parallel cases do you realistically have?"
Not that many really, basically the beta/dev testing for Kiloheartz and Stage Craft (maybe a dozen plugins). I will have a go at the rename feature and see how that pans out. I don't want to redo renaming each time I install a new version of the beta though (if so linvstmanager is not for me, and I'll do it manually).

**Comment by Goli4thus** [comment]
> Not that many really, basically the beta/dev testing for Kiloheartz and Stage Craft (maybe a dozen plugins). I will have a go at the rename feature and see how that pans out. I don't want to redo renaming each time I install a new version of the beta though (if so linvstmanager is not for me, and I'll do it manually).

If the path of the VST-dll stays the same accross different beta/dev-installs, then the renaming should be preserved (only the accompaning so-file besides the dll-file will have disappeared, which will make it display in LinVstManager as "No_So", which itself can be fixed via a simple "Update" action).
I don't know, in what way those beta/dev versions are being provided to you, but as long as you can adopt the install path to stay the same, this should work.

If on the other hand the path of the VST-dll changes between each beta/dev-install, then LinVstManager won't be able to locate it anymore and display "NotFound". You'd basically would need to re-add and re-rename the newly installed VST beta/dev-version.
So in that case Id say it's really up to what you think is more productive: 

- a) Re-add/re-scan for the newly installed VST with the additional renaming
- b) Doing it manually or with a custom script

In case you'd often be provided with a new beta/dev version in form of a "bundle install" (meaning you'd oftentimes reinstall all VSTs of a vendor), then maybe it would be helpful to have a "batch rename" feature.
I'm thinking of something like batch appending (or optionally prepending) a certain phrase the user enters.
This might actually come in handy if you'd do it like this:
1) Get a new beta/dev-installer
2) Install in parallel to previous versions
3) Add all to LinVstManager
4) Select all newly added VST3 (easily achieved with `View -> Filter`)
5) **Batch rename append** "_vst3" for instance
6) Now select all newly added VSTs of that specific beta/dev-release
7) **Batch rename prepend** "XYZ_", where XYZ is the version number of that release

That way you could have multiple versions installed in parallel and Bitwig would allow you to filter for a specific release in it's FX browser.

Batch renaming like this isn't a feature yet, but I think it'd be a useful one and could help in a beta testing scenario.
Anyway, if you think this might help you, I'd try to implement such a feature.

**Comment by keybreak** [comment]
> I'm thinking of something like batch appending (or optionally prepending) a certain phrase the user enters.

That's interesting idea for an advanced use cases!

Could be used very rarely by average users...But still nice to have if @perlindgren case has different plugin names for each beta build :smiley:

**Comment by perlindgren** [comment]
The plugins typically don't change names, and I usually just replace them, but in cases I have used a different install folder to be able to compare the versions. However those cases have been under pure windows, so far I haven't done it under wine (and if I recall correctly, I ran into some trouble with the DAW not liking two plugins with the same name, but that was under Live). So I was kind of hoping the situation would be better under wine + Bitwig. I'm currently building a new machine so can't test right now, but I hope I'll have it up and running shortly.

**Comment by Goli4thus** [comment]
Batch rename feature has been added: 5b3a1c6
Maybe it comes in handy for someone.

**Comment by keybreak** [comment]
@Goli4thus 
Tested, works like charm, haven't break anything.

**Comment by Goli4thus** [comment]
@perlindgren Clsoing this one for now. If it is still an issue, feel free to re-open it.
